### DISCUSSION TITLE
Including relationships in responses

### DISCUSSION THREAD

**Discussion by CEB-21** [tidal-music/discussions#47] [Ideas & Feature requests]
It'd be really nice to get tracks, artists, etc from a single response. So for instance, with Apple Music I can do something like:

    "albums/#{apple_id}?include[songs]=artists,albums&include[albums]=artists"

For an album lookup, this would give me back the artists and albums of that song and also gives me the album artists of that album. In doing so it could definitely cut back on the amount of requests. Alternatively it would be even better to get all of this in a single response without url params. Spotify and Deezer does this quite well.

**Comment by aliaksei-taliuk** [comment]
Hi @CEB-21. Welcome to the forum and thanks for your feedback. We're currently having the internal discussions about how to approach this and we're on it. Stay tuned!

**Comment by stefanmoro** [comment]
Hey @CEB-21,

A few months ago we started rolling out our "V2" API ([catalogue](https://developer.tidal.com/apiref?spec=catalogue-v2), [search](https://developer.tidal.com/apiref?spec=search-v2), [user](https://developer.tidal.com/apiref?spec=user-v2), [playlists](https://developer.tidal.com/apiref?spec=user-playlist-v2).), an improved version where we especially adress relationship handling. Our V2 API (which will be our only API going forward) does support relationship inclusion.

Thank you for your input, and hopefully this improvement addresses your use case!

**Reply by CEB-21** [reply]
This is wonderful news! Thanks so much!

**Reply by CEB-21** [reply]
Is there any way you can get any additional data from the includes besides `type` and `id`?

**Reply by stefanmoro** [reply]
You do get the full resource objects in the response as well - you find them in the `included` section.

**Reply by duttonst** [reply]
Hi. I've just started looking at reading playlists through the v2 API. Am I right in understanding that I can only specify `include=items` on a playlist query, not `include=albums` or `include=artists`? I've tried adding those 2 relationships but it doesn't seem to add any data to the `included` section of my response.

It would save a lot of additional queries if those were available: I would have thought most use cases involving tracks would as a minimum need artist name(s) as well.

**Reply by stefanmoro** [reply]
Hi @duttonst ,

It is correct that the API currently supports one level when [including related resources](https://jsonapi.org/format/#fetching-includes). Adding artists or albums as includes to a playlist wouldn't work, since a playlist has no such relationships.

For your use case, where I assume you want to iterate over a playlist and fetch the item and related information for each item, I would suggest the following:

Page through the `items` relationship for a playlist (TIDAL API defaults to a fixed global page size of 20). For each page, make a call to [get multiple tracks endpoint](https://developer.tidal.com/apiref?spec=catalogue-v2&ref=get-tracks-v2) for all items in that page, supplying the `include=albums,artists`.

That way you'd be able to fetch all info you need for all items in a page with 2 calls.

**Reply by duttonst** [reply]
Thank you @stefanmoro - that’s very helpful: how you describe is how I had ended up doing it so good to know I’m on the right track. 

However the implications of this are that reading a moderately long (1,500 track) playlist will need 150 queries. To put that in context, Spotify would need 16 queries to read a similar length playlist and Apple Music would need 15.

Reading that playlist on Apple Music, Spotify, Deezer or YouTube would take between 10 and 20 seconds. While I appreciate that your API is currently in development so performance comparisons are unfair, it’s currently taking more than 6 minutes to read the above playlist on Tidal. For use cases such as mine where I need the full playlist before I can start analysis of the contents this will be an issue. 

I hope you won’t mind me making the following observations regarding how other services achieve this level of performance:
1. Including data for “nested” objects is the norm. For example Apple Music allows `include[songs]=albums` to include album data on the tracks in a playlist. 
2. Page sizes of 100 (not 20) are the norm.
3. Throttling (429 errors) on playlist queries are quite a rare event (it’s common on catalog reads). I’m seeing a lot of 429s from your service when reading playlists, although of course that may well be because it’s not yet in production and because it’s taking so many queries to fully read a playlist in the first place. 

Again, I hope you don’t mind me making these observations - appreciate you are still in development but that’s usually the best time to address these kinds of issue so I thought they would be worth flagging.

**Reply by stefanmoro** [reply]
Thank you @duttonst, very helpful!

Point 1 and 2 are topics we have been, and still are, discussing internally. They also relate to the performance comment. For example, if we allow for nested includes and increase the page size, we would drastically reduce the number of roundtrips needed to our API in order to fetch the entire list of playlist items.

Regarding throttling, we currently apply an API wide access quota, which is quite strict now that we still are under development.

> Again, I hope you don’t mind me making these observations - appreciate you are still in development but that’s usually the best time to address these kinds of issue so I thought they would be worth flagging.

Not at all, we really appreciate your input. It helps us better understand how our API is used, not only by our own clients, but also for other use cases.

Thanks

**Reply by duttonst** [reply]
> Thank you @duttonst, very helpful!
> 
> Point 1 and 2 are topics we have been, and still are, discussing internally. They also relate to the performance comment. For example, if we allow for nested includes and increase the page size, we would drastically reduce the number of roundtrips needed to our API in order to fetch the entire list of playlist items.
> 
> Regarding throttling, we currently apply an API wide access quota, which is quite strict now that we still are under development.
> 
> > Again, I hope you don’t mind me making these observations - appreciate you are still in development but that’s usually the best time to address these kinds of issue so I thought they would be worth flagging.
> 
> Not at all, we really appreciate your input. It helps us better understand how our API is used, not only by our own clients, but also for other use cases.
> 
> Thanks

Hi Tidal.

Is there any news on this?

For us, resolving these performance bottlenecks would be a pre-requisite for going to production. However any news you have on when you might start accepting applications for production mode would also be much appreciated.

Thank you!

**Comment by patagonaa** [comment]
Sad to say this is still an issue and this, in combination with other factors (unnecessarily deeply nested JSON structure, aggressive rate limit) makes this API annoying at best and almost unusable at worst.

In my use case I need to get a playlist's items (including title and artist(s)) (simple enough).  
If, for example I want to do this with a (mind you, not very large) playlist of 250 songs, it's a matter of:

- request page (20 items) of playlist (`playlists/<id>/relationships/items`)
  - read `.data[].id` to get the track IDs
- request tracks and artists for page (`tracks?include=artists&filter[id]=1111&filter[id]=2222[...]`)
- Then I have 3 separate lists: playlist items, tracks, artists
  - for each playlist item, I have to lookup the track (so I can get the name from `.attributes.name` and its artist id from `.relationships.artists.data[].id`)
    - for each artist id, I have to lookup the artist id in `.included[]` to access `.attributes.name`
- Then, repeat until `.links.next` is empty

Which means I need to do 26 requests (which runs into the rate limit very quickly) and juggle 3 lists (playlist items, tracks, artists) in my code instead of just getting all the data nested into each other.  
I do understand that it makes sense to keep these entities normalized in a database, but in an API it just makes things difficult, _especially for these kinds of entities that do not make sense without its related ones_ (how useful is a playlist item without the track? how useful is a track without its artist?).

**Reply by stefanmoro** [reply]
Hi @patagonaa, thanks for reaching out.

If I understand your case correctly, you should be able to achieve what you want (getting track and artist(s) infromation for items in a playlist, and iterate over that) in one request per page: 
`GET playlists/<id>/relationships/items?countryCode=US&include=items.artists`

And paginate through that. Refs linkage, there are several client side libraries that should help with that - some [listed here](https://jsonapi.org/implementations/#client-libraries).

Hope that helps.

**Reply by patagonaa** [reply]
wow, that actually worked.

Reading the OpenAPI spec and the above comments, it sounded like it wasn't possible to include nested items like that, only one "additional" level per request, especially after you said

> the API currently supports one level when [including related resources](https://jsonapi.org/format/#fetching-includes)

Has that changed (and wasn't reflected well in the docs) or was that just worded in an unclear way (to me)?

Other than the other minor annoyances (that seem to come from JSON API mostly) and the rate limiting (which is probably bearable now that I have to do only half the requests), this solves my issues.

Thank you!

**Reply by patagonaa** [reply]
@stefanmoro
I think I found a small bug with that:  
When doing the request with `include=items.artists`, the next page link in the response only has `include=items`, so on the next page, the artists are missing again.

This can be worked around by changing the URL before requesting the next page, but is still a bug imo.

**Reply by patagonaa** [reply]
Seems to have been fixed some time between now and my last comment (`next` URL now properly has the entire include path). Thanks!
